src/tsc/tsc.ts

```diff
@@ -9,16 +9,39 @@ ts.Debug.loggingHost = {
     },
 };
 
-if (ts.Debug.isDebugging) {
-    ts.Debug.enableDebugInfo();
+let v8: typeof import("v8") | undefined;
+try {
+    v8 = require("v8");
 }
-
-if (ts.sys.tryEnableSourceMapsForHost && /^development$/i.test(ts.sys.getEnvironmentVariable("NODE_ENV"))) {
-    ts.sys.tryEnableSourceMapsForHost();
+catch {
+    // do nothing
 }
 
-if (ts.sys.setBlocking) {
-    ts.sys.setBlocking();
+function main() {
+    if (ts.Debug.isDebugging) {
+        ts.Debug.enableDebugInfo();
+    }
+
+    if (ts.sys.tryEnableSourceMapsForHost && /^development$/i.test(ts.sys.getEnvironmentVariable("NODE_ENV"))) {
+        ts.sys.tryEnableSourceMapsForHost();
+    }
+
+    if (ts.sys.setBlocking) {
+        ts.sys.setBlocking();
+    }
+
+    ts.executeCommandLine(ts.sys, ts.noop, ts.sys.args);
 }
 
-ts.executeCommandLine(ts.sys, ts.noop, ts.sys.args);
+if ((v8 as any)?.startupSnapshot.isBuildingSnapshot()) {
+    (v8 as any).startupSnapshot.setDeserializeMainFunction(() => {

```

Comment:
Likely what I should be doing here (and in sys.ts) is instead use `addDeserializeCallback` to hook calls that fix up existing stuff. That way, it can work for all of our entrypoints. I can see snapshotting benefiting tsserver startup time, for example, though that's a little less critical given it's long running.

A gotcha is that we will need to think about people importing TypeScript and snapshotting themselves; theoretically we don't care because the top-level `ts.sys` should not be mutable in public.